### PR TITLE
Fix Consent API integration with OSA

### PR DESCRIPTION
**Pull request by tomalec**
This is a follow-up from https://github.com/woocommerce/woocommerce/pull/39701
that requires https://github.com/woocommerce/woocommerce/pull/41690
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

- [Stop tracking order attribution when the consent is revoked on the same page.](https://github.com/woocommerce/woocommerce/commit/60d16fe9f6a83836830c83c19352144ae147b8a4) 
    Switch `wp_consent_type_defined` to VanillaJS event, as this is what the latest Complianz dispatches.
    Remove jQuery dependency.
- [Update checkout block data when consent is given/revoked on the same page](https://github.com/woocommerce/woocommerce/pull/41856/commits/c2deff5176d9d44671ea4a09301f9ea116a33aac) 
c2deff5

Closes # .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Setup another plugin that uses WP Consent API, like Complianz

##### Denied by default
When you install Compilanz, the marketing data collection should be disabled by default, so even if you do not click anything in the cookie banner, `wc_order_attribution.params.allowTracking` should be set to `false` and no order attribution data should be collected. 

:lady_beetle: This integration was broken as the `WPConsentAPI` register was not called, and `-blocs` integration disregarded `wc_order_attribution.params.allowTracking` setting.

1. Open shop page in incognito mode. Do not click anything in the cookie banner
2. Add something to the cart and proceed to the block/classic checkout
3. Open dev tools
   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have empty values
   - For block checkout, monitor network tab and `/checkout` XHR
   - `wc_order_attribution.params.allowTracking` should be set to `false` 
4. "Place order"
	- For classic checkout, confirm that submitted POST **does not** contain `wc_order_attribution_*`  values
	- For block checkout, the `woocommerce/order-attribution` extension payload contains all `null`s 
6. Check that the placed order has no order attribution data in the `/wp-admin/admin.php?page=wc-orders&action=edit&id=`

##### Consent granted on the checkout page
:bug: This behavior was covered by JS code for classic checkout, so only registering `WPConsentAPI` would be all we need. But the behavior was broken for blocks

1. Start a new incognito session. Do not click anything in the cookie banner
2. Add something to the cart and proceed to the block/classic checkout
3. Open dev tools
   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have empty values
   - For block checkout, monitor network tab and `/checkout` XHR
   - `wc_order_attribution.params.allowTracking` should be set to `false` 
4. Grant "marketing" consent without reloading the page
   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have set values
   - `wc_order_attribution.params.allowTracking` should be set to `true` 
4. "Place order"
	- For classic checkout, confirm that submitted POST **does** contain `wc_order_attribution_*`  values
	- For block checkout, the `woocommerce/order-attribution` extension payload contains OSA data
6. Check that the placed order has order attribution data in the `/wp-admin/admin.php?page=wc-orders&action=edit&id=`

##### Consent granted before
1. Start a new incognito session. Do grant marketing consent in the cookie banner, reload the page
2. Add something to the cart and proceed to the block/classic checkout
3. Open dev tools
   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have set  values
   - For block checkout, monitor the network tab and `/checkout` XHRvalues
   - `wc_order_attribution.params.allowTracking` should be set to `true` 
4. "Place order"
	- For classic checkout, confirm that submitted POST **does** contain `wc_order_attribution_*`  values
	- For block checkout, the `woocommerce/order-attribution` extension payload contains OSA data
6. Check that the placed order has order attribution data in the `/wp-admin/admin.php?page=wc-orders&action=edit&id=`

##### Consent revoked on the checkout page
I didn't manage to simulate this behavior with the Compilanz UI, but theoretically, API supports this, so I believe we should be prepared for such behavior from another plugin.

To support this behavior, we need the majority of JS changes from this PR.

1. Start a new incognito session. Do grant marketing consent in the cookie banner, reload the page, or play with granting it at any later step before revoking it.
2. Add something to the cart and proceed to the block/classic checkout
3. Open dev tools
   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have set values
   - For block checkout, monitor the network tab and `/checkout` XHRvalues
   - `wc_order_attribution.params.allowTracking` should be set to `true` 
4. :yellow_circle: Simulate revoking the consent 
   ```js
   document.body.dispatchEvent(new CustomEvent('wp_listen_for_consent_change',{detail: {marketing: 'no'}}));
   ```

   - For classic checkout: Check that all `document.querySelectorAll('input[name^=wc_order_attribution_]');` inputs have empty values
   - `wc_order_attribution.params.allowTracking` should be set to `false` 
4. "Place order"
	- For classic checkout, confirm that submitted POST **does not** contain `wc_order_attribution_*`  values
	- For block checkout, the `woocommerce/order-attribution` extension payload contains all `null`s
6. Check that the placed order has no order attribution data in the `/wp-admin/admin.php?page=wc-orders&action=edit&id=`

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Comment by tomalec** [comment]
as far as I tested in `trunk` the `WPConsentAPI` is not registered when Complianz is setup, so the integration js script is not loaded at all.

File: plugins/woocommerce/src/Internal/Orders/OrderAttributionBlocksController.php [hunk 1]
```diff
@@ -72,50 +72,6 @@ public function register() {
...
-	 * Register scripts.
-	 */
-	private function register_assets() {
-		wp_register_script(
-			'wc-order-attribution-blocks',
-			plugins_url(
-				"assets/js/frontend/order-attribution-blocks{$this->get_script_suffix()}.js",
-				WC_PLUGIN_FILE
-			),
-			array( 'wc-order-attribution', 'wp-data', 'wc-blocks-checkout' ),
```
Comment (tomalec) [inline plugins/woocommerce/src/Internal/Orders/OrderAttributionBlocksController.php:106]: This is a bit tricky bit. I'm not sure how to handle it.
as the `wc-order-attribution`  script now requires `'wp-data', 'wc-blocks-checkout'` dependencies, but only for Block-based checkout, and does not require it for the classic one.

Can we add them conditionally? Or do we **always**(?) load block dependencies in the classic checkout anyway?

Comment (message-dimke) [reply plugins/woocommerce/src/Internal/Orders/OrderAttributionBlocksController.php:106]: From what I debugged both Attribution Controllers load no matter which checkout is used, so all the scrips would load anyways and I do not think it is possible for backend to know which layout we are using on the front at this stage. This is the case when resolving this issue will take 80% of time but less than 20% of income. Maybe we leave it like this for now?

**Comment by message-dimke** [comment]
### Intro.

I did test the PR and I think it works as described. Had to spend some time on configuring Complianz though (somehow did configure it to allow everything by default so was testing wise versa).

#### Disallowing tracking did its thing.

Fullscreen_08_12_2023__11_22

and sent nothing to the backend

Cursor_and_DevTools_-_wordpress_ddev_site_checkout_order-received_35938__key_wc_order_Mnx1sMr2ECKLB


so Orders did not have anything tracked

Orders_‹_Wordpress_—_WordPress


The same is true for the Checkout Block

Fullscreen_09_12_2023__19_04

#### I did also try to simulate consent change during the checkout flow.

...by manually triggering 'marketing' consent with 

```javascript
// Marketing allow
document.dispatchEvent(new CustomEvent('wp_listen_for_consent_change',{detail: {marketing: 'allow'}}));
// Marketing dissallow
document.dispatchEvent(new CustomEvent('wp_listen_for_consent_change',{detail: {marketing: 'no'}}));
```

and things also did work out.

Fullscreen_09_12_2023__19_59

Overall I think the code works as expected and is good to go.

Regarding doing `WPConsentAPI::register` inside Attribute Controller class register method, I, personally, think it is fine since we need consent api for order attribution only, at least for now.

**Review by message-dimke — APPROVED** [review]
Looks good and works as expected!

Thank you, @tomalec !

**Comment by layoutd** [comment]
Hey @tomalec, can you clarify the difference between this and #43012? 😃 
Should we still be pushing both towards review+merge? Or can this one be closed in favor of the other PR?

**Comment by tomalec** [comment]
The difference is in the target branch. 
We should merge one or the other. Depending on https://github.com/woocommerce/woocommerce/pull/41690

If https://github.com/woocommerce/woocommerce/pull/41690 gets merged first, we should push this PR.
If not, we should push for https://github.com/woocommerce/woocommerce/pull/43012.
Then, once #43012 is merged, we should close this one and update #41690 to accommodate new changes.

**Comment by tomalec** [comment]
Closing in favor of https://github.com/woocommerce/woocommerce/pull/43012

**Pull request closed without merging**